app/controllers/api/v1/service_plans_controller.rb

```diff
@@ -6,21 +8,59 @@ class ServicePlansController < ApplicationController
 
       def order
         service_plan = model.find(params_for_order[:service_plan_id].to_i)
-        task = Task.create!(:tenant => service_plan.tenant, :state => "pending", :status => "ok")
 
-        messaging_client.publish_message(
-          :service => "platform.topological-inventory.operations-openshift",
-          :message => "ServicePlan.order",
-          :payload => payload_for_order(task, service_plan)
-        )
+        retrieve_source_type(service_plan) do |source_type|
+          task = Task.create!(:tenant => service_plan.tenant, :state => "pending", :status => "ok")
+
+          messaging_client.publish_message(
+            :service => "platform.topological-inventory.operations-#{source_type.name}",
+            :message => "ServicePlan.order",
+            :payload => payload_for_order(task, service_plan)
+          )
+
+          render :json => {:task_id => task.id}
+        end
 
-        render :json => {:task_id => task.id}
       rescue ActiveRecord::RecordNotFound
         head :bad_request
+      rescue StandardError => err
+        error_document = ManageIQ::API::Common::ErrorDocument.new.add(err.message)
+        render :json => error_document.to_h, :status => error_document.status
       end
 
       private
 
+      def sources_api_client
+        @sources_api_client ||=
+          begin
+            identity = { 'x-rh-identity' => request.headers.fetch('x-rh-identity') }
+            api_client = SourcesApiClient::ApiClient.new
+            api_client.default_headers.merge!(identity)
+            SourcesApiClient::DefaultApi.new(api_client)
+          end
+      end
+
+      def retrieve_source_type(service_plan)
+        source = sources_api_client.show_source(service_plan.source_id.to_s)
+        if source.present?
+          source_type = sources_api_client.show_source_type(source.source_type_id.to_s)
+          if source_type.present?
+            #
+            # Process Source Type
+            #
+            yield source_type
```

Comment:
I don't think the yield/block buys you anything besides increased indentation, just raise if the source_type is nil then return it